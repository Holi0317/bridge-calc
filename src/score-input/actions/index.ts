export * from "./bid";
export * from "./change-players";
export * from "./set-bid";
export * from "./set-win";
export * from "./skip";
export * from "./start";
export * from "./undo";
export * from "./win";
export * from "./replace-current-game";
