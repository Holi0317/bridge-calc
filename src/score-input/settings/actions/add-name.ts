import {ActionTypes} from '../../../action-types'
import * as cuid from 'cuid'
import {randomName} from '../../../example-names'

export interface IAddNameAction {
  type: ActionTypes.ADD_NAME,
  /**
   * Name of the new player
   */
  name: string,
  /**
   * A new ID for him/her.
   */
  ID: string
}

/**
 * Add a name to settings UI state and append to the end of the list.
 * This does NOT change currentGame state in the store.
 * @param name - Name of the new Player
 */
export function addNameAction(name: string): IAddNameAction {
  return {type: ActionTypes.ADD_NAME, name, ID: cuid()}
}

export function addRandomNameAction(): IAddNameAction {
  return {type: ActionTypes.ADD_NAME, name: randomName(), ID: cuid()}
}
