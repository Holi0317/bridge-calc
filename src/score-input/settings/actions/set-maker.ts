import { ActionTypes } from "../../../action-types";

export interface ISetMakerAction {
  type: ActionTypes.SET_MAKER;
  /**
   * The Player ID of maker to be set.
   */
  maker: string;
}

/**
 * Set maker in UI settings UI.
 * This does NOT change anything in currentGame state of our store.
 * @param maker - Player ID of the maker to be set.
 */
export function setMakerAction(maker: string): ISetMakerAction {
  return { type: ActionTypes.SET_MAKER, maker };
}
