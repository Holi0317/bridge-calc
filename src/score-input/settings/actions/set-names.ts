import fromPairs from 'lodash-es/fromPairs'
import {ActionTypes} from '../../../action-types'
import {IPlayerMap} from '../../../types'
import {PlayerName} from '../selectors/name-input-list-source'

export interface ISetNamesAction {
  type: ActionTypes.SET_NAMES,
  /**
   * New name map to be set.
   */
  newNames: IPlayerMap<string>
}

/**
 * Set names state in settings UI.
 * This does NOT change anything in currentGame state in the store
 * @param newNames - New name map to be set.
 */
export function setNamesAction(newNames: IPlayerMap<string>): ISetNamesAction {
  return {type: ActionTypes.SET_NAMES, newNames}
}

export function setNamesFromArrayAction(newNames: PlayerName[]): ISetNamesAction {
  return {type: ActionTypes.SET_NAMES, newNames: fromPairs(newNames)}
}
