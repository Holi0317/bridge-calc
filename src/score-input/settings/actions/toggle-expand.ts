import { PANEL } from "../panel";
import { ActionTypes } from "../../../action-types";

export interface ToggleExpandAction {
  type: ActionTypes.TOGGLE_SETTING_PANEL;
  /**
   * The panel to be toggled.
   */
  panel: PANEL;
}

/**
 * Toggle expansion state a setting panel
 *
 * @param panel - The panel to be toggled
 */
export function toggleExpandAction(panel: PANEL): ToggleExpandAction {
  return { type: ActionTypes.TOGGLE_SETTING_PANEL, panel };
}
