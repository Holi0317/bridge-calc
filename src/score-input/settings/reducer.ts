import { GameStage } from "../game-stage";
import { PlayerMap } from "../../types";
import { ActionTypes } from "../../action-types";
import { PANEL } from "./panel";
import { createReducer } from "typesafe-actions";

export interface SettingsState {
  /**
   * Player ID of currently selected maker.
   * If null, no maker is selected or the game is not running.
   */
  readonly maker: string | null;
  /**
   * If maker has changed by user or not.
   * By default, this value is false.
   */
  readonly makerDirty: boolean;
  /**
   * New player map.
   */
  readonly names: PlayerMap<string>;
  /**
   * State which panel is expanded now.
   * If null, no panel is expanded.
   */
  readonly expandedPanel: PANEL | null;
}

const defaultState: SettingsState = {
  maker: null,
  makerDirty: false,
  names: {},
  expandedPanel: null
};

export const settingsReducer = createReducer(defaultState)
  .handleType(ActionTypes.INIT_SETTINGS, (_, { payload }) =>
    payload == null || payload.stage === GameStage.ended
      ? defaultState
      : {
          ...defaultState,
          maker: payload.currentPlayerOrder[0],
          names: payload.names
        }
  )
  .handleType(ActionTypes.SET_MAKER, (state, { payload }) => ({
    ...state,
    makerDirty: true,
    maker: payload
  }))
  .handleType(ActionTypes.SET_NAMES, (state, { payload }) => ({
    ...state,
    names: payload
  }))
  .handleType(ActionTypes.ADD_NAME, (state, { payload }) => ({
    ...state,
    names: {
      ...state.names,
      [payload.id]: payload.name
    }
  }))
  .handleType(ActionTypes.TOGGLE_SETTING_PANEL, (state, { payload }) => ({
    ...state,
    expandedPanel: payload === state.expandedPanel ? null : payload
  }));
