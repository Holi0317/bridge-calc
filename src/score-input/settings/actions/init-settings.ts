import { ActionTypes } from "../../../action-types";
import { GameState } from "../../reducer";

export interface IInitSettingsAction {
  type: ActionTypes.INIT_SETTINGS;
  /**
   * Game state to be followed for setting settings state.
   */
  state: GameState;
}

/**
 * Initiate state of settings section by given game state.
 * If the state is null or stage is end, settings state will return to default.
 * @param state - Game state of on-going session
 */
export function initSettingsAction(state: GameState): IInitSettingsAction {
  return { type: ActionTypes.INIT_SETTINGS, state };
}
