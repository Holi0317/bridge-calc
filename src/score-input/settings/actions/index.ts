import { InitSettingsAction } from "./init-settings";
import { SetMakerAction } from "./set-maker";
import { SetNamesAction } from "./set-names";
import { AddNameAction } from "./add-name";
import { ToggleExpandAction } from "./toggle-expand";

export type GameSettingsActions =
  | InitSettingsAction
  | SetMakerAction
  | SetNamesAction
  | AddNameAction
  | ToggleExpandAction;
