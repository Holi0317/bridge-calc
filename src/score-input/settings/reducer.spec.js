import {settingsReducer as reducer} from './reducer'
import {defaultState} from '../../../test-fixtures/settings-state'
import {endedState, genMap, waitingBidState} from '../../../test-fixtures/current-game-states'
import {initSettingsAction} from './actions/init-settings'
import {setMakerAction} from './actions/set-maker'
import {setNamesAction} from './actions/set-names'
import {addNameAction} from './actions/add-name'

jest.mock('cuid', () => {
  let count = 0
  return jest.fn(() => '' + count++)
})

test('Default state', () => {
  const state = undefined
  const expected = {
    ...defaultState
  }
  const action = {
    type: '#NULL'
  }
  const actual = reducer(state, action)
  expect(actual).toEqual(expected)
})

test('INIT_SETTINGS should return default state on null game state', () => {
  const state = null
  const expected = {
    ...defaultState
  }
  const action = initSettingsAction(null)
  const actual = reducer(state, action)
  expect(actual).toEqual(expected)
})

test('INIT_SETTINGS should return default state on ended game stage', () => {
  const state = null
  const expected = {
    ...defaultState
  }
  const action = initSettingsAction(endedState)
  const actual = reducer(state, action)
  expect(actual).toEqual(expected)
})

test('INIT_SETTINGS should work on running game state', () => {
  const state = null
  const expected = {
    maker: 'a',
    names: genMap('John', 'Mary', 'Henry', 'Joe')
  }
  const action = initSettingsAction(waitingBidState)
  const actual = reducer(state, action)
  expect(actual).toEqual(expected)
})

test('SET_MAKER should set maker', () => {
  const state = {
    ...defaultState
  }
  const expected = {
    ...defaultState,
    maker: 'b'
  }
  const action = setMakerAction('b')
  const actual = reducer(state, action)
  expect(actual).toEqual(expected)
})

test('SET_NAMES should set names', () => {
  const names = genMap('Mary', 'John', 'Henry', 'Joe')
  const state = {
    ...defaultState
  }
  const expected = {
    ...defaultState,
    names
  }
  const action = setNamesAction(names)
  const actual = reducer(state, action)
  expect(actual).toEqual(expected)
})

test('ADD_NAME should append new name to the end', () => {
  const state = {
    ...defaultState
  }
  const expected = {
    ...defaultState,
    names: {
      0: 'DPGJW'
    }
  }
  const action = addNameAction('DPGJW')
  const actual = reducer(state, action)
  expect(actual).toEqual(expected)
})
