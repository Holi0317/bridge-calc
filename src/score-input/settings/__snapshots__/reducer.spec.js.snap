// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`INIT_SETTINGS should work on running game state 1`] = `
Object {
  "expandedPanel": null,
  "maker": "a",
  "makerDirty": false,
  "names": Object {
    "a": "John",
    "b": "Mary",
    "c": "Henry",
    "d": "Joe",
  },
}
`;
